### ISSUE TITLE
High memory consumption when running openie in simple.Document because of internal cache

### ISSUE THREAD

**Issue by renatodinhani** [stanfordnlp/CoreNLP#151]
Class `edu.stanford.nlp.simple.Document` seems to be keeping sort of an internal cache in the attributes `customAnnotators`, `annotationPoolKeys` and `annotationPool`.

Because `openie` method consumes a significant ammount of memory, memory is filled up to 4GB with just 10 documents processed.

I think would be nice if we can disable the cache, limit it or provide a method to clear it. Currently I am clearing the cache using reflection after each document is processed with the following code and everything still working:

```
FieldUtils::readStaticField(self.getClass(), "customAnnotators", true).clear()
FieldUtils::readStaticField(self.getClass(), "annotationPoolKeys", true).clear()
FieldUtils::readStaticField(self.getClass(), "annotationPool", true).clear()
```

**Comment by gangeli** [comment]
Can you post some code I can run to debug this? In theory this shouldn't happen (which of course doesn't mean it doesn't) -- there are at least two blocks against it:

(1) the custom annotator pool checks if there's already an annotator of the given signature defined: https://github.com/stanfordnlp/CoreNLP/blob/master/src/edu/stanford/nlp/simple/Document.java#L240

(2) No more than 10 custom annotators can be created at a time: https://github.com/stanfordnlp/CoreNLP/blob/master/src/edu/stanford/nlp/simple/Document.java#L246

**Comment by renatodinhani** [comment]
I am not sure if the problem are the `Annotators`. In my tests I was monitoring the `Document` instances and they were not garbage collected before I mannually cleared that internal attributes.

The following lines tell me this list is holding until 64 documents in memory.

https://github.com/stanfordnlp/CoreNLP/blob/master/src/edu/stanford/nlp/simple/Document.java#L845-L847

 If I don't run `openie()`, I can observe that behavior in action, they will be removed after some time. The problem is when running `openie()`, memory will fill fast before reaching 64.

About the code sample, I am running with JRuby. Here is a simplified version of the code. It does not do anything complex, just `lemmas()` and `openie()`.

```
module NLP
  include_package 'edu.stanford.nlp.simple'  

  class Processor

    def process(input_text)
      output_text = ""
      informations = []

doc = JDocument.new(content)

      # lemmatize text and extract informations
      doc.sentences.each do |sentence|
        output_text += sentence.lemmas().join(" ") + "\n"
        informations << sentence.openie()
      end
      informations = informations.flatten()

      # return values
      return output_text, informations
    end
  end
end
```

**Comment by renatodinhani** [comment]
I am not sure if the problem are the `Annotators`. In my tests I was monitoring the `Document `instances and they were not garbage collected before I mannually cleared that internal attributes.

The following lines tell me this list is holding until 64 documents in memory.

https://github.com/stanfordnlp/CoreNLP/blob/master/src/edu/stanford/nlp/simple/Document.java#L845-L847

If I don't run `openie()`, I can observe that behavior in action, they will be removed after some time. The problem is when running `openie()`, memory will fill fast before reaching 64.

About the code sample, I am running with JRuby. Here is a simplified version of the code. It does not do anything complex, just `lemmas()` and `openie()`.

```ruby
module NLP
  include_package 'edu.stanford.nlp.simple'  

  class Processor

    def process(input_text)
      output_text = ""
      informations = []

      # create Document 
      doc = Document.new(input_text)

      # lemmatize text and extract informations
      doc.sentences.each do |sentence|
        output_text += sentence.lemmas().join(" ") + "\n"
        informations << sentence.openie()
      end
      informations = informations.flatten()

      # return values
      return output_text, informations
    end
  end
end
```

**Comment by gangeli** [comment]
Huh; this should certainly not break. I'll copy the code to replicate the bug from Java. How much memory are you giving the JVM, and how long is your longest sentence?

**Comment by renatodinhani** [comment]
I am giving 4GB to JVM. I am processing 23000 Wikipedia texts about games (without HTML tags) calling that `process` method for each document.

I did some tests using the `StandfordCoreNLP` class and it works well. For me it does not seems to be a problem with the algorithms itself, but something with the class `Document` preveting some data to be collected.

The following comment give me the hint to clear the static cache because that could be causing the problem:
https://github.com/stanfordnlp/CoreNLP/blob/master/src/edu/stanford/nlp/simple/Document.java#L252-L256

**Comment by gangeli** [comment]
Ah, I see. It surprises me that memory consumption increases so fast, but I guess our Annotation objects are bigger than we think...

I've pushed a hopeful fix -- rather than limiting the cache by number of documents, I'm trying to use SoftReferences to let the garbage collector decide when to free up memory. Let me know if it helps!

**Comment by gangeli** [comment]
OK, I got rid of this cache altogether. Seems like more of a pain than it's worth.

**Comment by mjball** [comment]
What's the ETA on a release that contains this fix? The [latest I see on Maven Central is 3.6.0, published in January 2016.](http://search.maven.org/#search%7Cga%7C1%7Cstanford-corenlp)

 I'd like to get this fix in my build sooner rather than later, as the old behavior is causing OOMs.

**Comment by mjball** [comment]
As an aside: the static `IdentityHashMap<Document, Annotation>` seemed unnecessarily complicated to me. Unless I'm missing something, the non-static `SoftReference<Annotation>` should be pretty much equivalent, without the nasty `finalize()`.
